### ISSUE TITLE
Enhancement: [no-unnecessary-type-arguments] don't report explicit type argument if it's an object shape

### ISSUE THREAD

**Issue by ulrichstark** [typescript-eslint/typescript-eslint#12160]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-type-arguments

### Description

I would like to have the rule not report on some explicit type arguments (maybe all type arguments with object shape). In its current state, the rule reports and wants to remove type arguments, that help me keep potential TS errors close to where they need to be fixed.

My example code and repro below illustrate the problem. I have an useState of a specific type. The initial value of the useState makes the type argument unnecessary/redundant, so the logic of the rule is technically correct. But this pattern helps me find the location where I need to change code, when the interface Values changes in the future. If the type argument wasn't explicitly declared, the error would surface in other irrelevant code locations or even not at all.

But honestly I'm not sure how the logic of the rule should adapt to that and if that makes more sense as a rule option or not.

### Fail

```TypeScript
--
```

### Pass

```TypeScript
interface Values {
  key: string;
}

declare function useState<S>(initialState: S): [S, () => S];

const [values, setValues] = useState<Values>({
  key: "value"
});
```

### Additional Info

Repro: https://typescript-eslint.io/play#ts=5.9.3&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgGpwDYFcIGdkDeAUMsgNYQCeAXMrmFKAOYDcRAvkUQCYQIZwoKGFhAIwwAPYhkWXBADKYOJAA8CgHwAKUMAmYlKiLQUBKWgG0FAGmRbTyALwbkCgLpsiCafWQWAbpg4uLbyYOjYeG5OsvKGahHB2sSkFDTIAESBkRkcpixAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDJETRoHaJHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by bradzacher** [comment]
On the surface at least this neuters the entire rule to almost nothing.
it'll only catch cases where you explicitly pass a primitive that might have been inferred -- which seems like a very very rare case. 

it feels like it would be better to just disable the rule entirely rather than burning the CPU time checking every single explicit type arg to see if it's a primitive that is unnecessary.
